#!/usr/bin/env python
PACKAGE = "vrmagic_device"

from dynamic_reconfigure.parameter_generator_catkin import *
from driver_base.msg import SensorLevels

gen = ParameterGenerator()

# Name, Type, Reconfiguration level, Description, Default, Min, Max

gen.add("frame_id", str_t, SensorLevels.RECONFIGURE_CLOSE,
        "ROS tf frame of reference, resolved with tf_prefix unless absolute.",
        "vrmagic")

gen.add("frame_rate", double_t, SensorLevels.RECONFIGURE_RUNNING,
        "Frame rate for all cameras (frames per second).", 13.0, 1.0, 60.0)

formats = gen.enum([gen.const("mono8", str_t, "mono8", ""),
                    gen.const("mono16", str_t, "mono16", ""),
                    gen.const("mono8rle", str_t, "mono8rle", ""),
                    ],
                   "Source (raw) format of the device.")

gen.add("source_format", str_t, SensorLevels.RECONFIGURE_CLOSE,
        "Source format.", "mono8",
        edit_method = formats)

gen.add("external_trigger", bool_t, SensorLevels.RECONFIGURE_RUNNING,
        "External trigger.", False)

gen.add("auto_exposure", bool_t, SensorLevels.RECONFIGURE_RUNNING,
        "Auto exposure.", True)

gen.add("exposure_time", double_t, SensorLevels.RECONFIGURE_RUNNING,
        "Exposure time (ms).", 10, 0.1, 500)

gen.add("auto_exposure_time_max", double_t, SensorLevels.RECONFIGURE_RUNNING,
        "Maximum auto exposure time (ms).", 40, 0.1, 75)

gen.add("auto_exposure_bright_threshold", int_t, SensorLevels.RECONFIGURE_RUNNING,
        "Bright threshold of pixels used for AEC/AGC.", 58, 1, 64)
        
gen.add("auto_exposure_bright_percentage", double_t, SensorLevels.RECONFIGURE_RUNNING,
        "Bright pixels target percentage for AEC/AGC.", 12, 0.0, 18)

gen.add("auto_gain", bool_t, SensorLevels.RECONFIGURE_RUNNING,
        "Auto gain.", True)

gen.add("gain", int_t, SensorLevels.RECONFIGURE_RUNNING,
        "Gain", 16, 16, 64)
        
gen.add("auto_gain_max", int_t, SensorLevels.RECONFIGURE_RUNNING,
        "Maximum auto gain.", 32, 16, 64)

gen.add("anti_blooming", bool_t, SensorLevels.RECONFIGURE_RUNNING,
        "Anti blooming.", True)
        
gen.add("anti_blooming_voltage", double_t, SensorLevels.RECONFIGURE_RUNNING,
        "Anti blooming voltage.", 1.95, 1.95, 2.3)
        
gen.add("high_dynamic_mode", bool_t, SensorLevels.RECONFIGURE_RUNNING,
        "High dynamic mode.", True)

exit(gen.generate(PACKAGE, "VRmagicDevice", "VRmagicDevice"))
